wip-mariocao-median: draft submission (#66)
witnet/WIPs: wip-mariocao-median.md

<pre>

>   WIP: WIP-0015

Reviewer: Please wait for a WIP number to be assigned

Reviewer 2: The WIP editor is not happy.

Author: Ups...that was a copy paste as it is already in use! :+1:

  Layer: Consensus (hard fork)
  Title: Add average median to RADON reducers
  Authors: Mario Cao <mario@witnet.foundation>
  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
  Status: Draft
  Type: Standards Track
  Created: 2020-09-01
  License: BSD-2-Clause
</pre>

## Abstract

The **average median** is proposed to be included in the reducers catalog available in the RAD engine. Data requests will then support median function as part of the **retrieval**, **aggregation** and **tally** scripts.

## Motivation and rationale

It is paramount to continue increasing the catalog of available reducer functions for data requests within the RAD engine to increase the flexibility and potential of supporting additional use cases by the Witnet protocol.

This document proposes to add the support of the **Median** as part of the `reducer` functions because, depending on the use case, it may be more descriptive than the average mean.

Specifically for the Witnet protocol, it is interesting because of the following reasons:

>  1. It mitigates the *negative* effect that outliers may have in a data request result at any stage where multiple data points are aggregated. For example, a data request using the **mean** function is sensitive to tampered with if malicious node reports are not filtered out previously.

Reviewer 2: ```suggestion
 1. It mitigates the *negative* effect that outliers may have in a data request result at any stage where multiple data points are aggregated. For example, a data request using the **mean** function is sensitive to be tampered with if malicious reports are not filtered out previously.
```

>  2. It can is advantageous with the commit/reveal scheme of the data request lifecycle. Malicious nodes cannot affect a result unless they take over most of the data request committee.

Reviewer 2: ```suggestion
 2. It can be advantageous with the commit/reveal scheme of the data request lifecycle. Malicious nodes cannot affect a result unless they take over most of the data request committee.
```

Reviewer 2: I don't really follow this argument. How is it different from the former?

Author: How about this? 

From a data request perspective, the median is interesting because it mitigates the *negative* effect that outliers may have in a data request result at any stage where multiple data points are aggregated. For example, a data request using the **mean** function is sensitive to be tampered with if malicious reports are not filtered out previously.

The median might be sensitive to "last reveal attacks" if at some point of the data aggregation, one of the witnessing nodes waits to know all other data points to influence the median result. However, this is not possible in the Witnet protocol as witnessing nodes follow a commit/reveal scheme.

## Specification

> This document proposes to add the support of the **Median** to the available RADON reducer functions.

Reviewer 2: No need to introduce the changes in the specification, which should be as straight to the point as possible. This belongs in the other sections before.

> The *Median* is defined as the "middle number" in a sorted, ascending or descending, list of numbers. If the data set has an odd number of observations, the number in the middle of the list is the median. Else, if data set has an even number of observations, then the median is defined as the arithmetic mean of the two middle values.

Reviewer 2: ```suggestion
The *Median* is defined as the "middle number" in a sorted, ascending or descending, list of numbers. If the data set has an odd number of observations, the number in the middle of the list is the median. Otherwise, if the data set has an even number of observations, then the median is defined as the arithmetic mean of the two middle values.
```

> **(1)**: The median will used the already defined identifier for `AverageMedian` as `0x05`.

Reviewer 2: ```suggestion
**(1)**: The median will use the already defined operator identifier for `AverageMedian`, which is `0x05`.
```

> **(2)**: The median reducer will only support the inputs of the types `RadonTypes::Float` and `RadonTypes::Integer`. 

Reviewer 2: `RadonTypes` is specific to witnet-rust
```suggestion
**(2)**: The median reducer will only support homogeneous arrays containing exclusively `Float` or `Integer` values. 
```

> **(3)**: `NaN` values are filtered out and ignored before sorting the list of observations. `Infinity` and `-Infinity` values are considered valid.

Reviewer 2: ```suggestion
**(3)**: `NaN` values are filtered out and ignored before sorting the array. `Infinity` and `-Infinity` values are considered valid.
```

> **(4)**: If the median is applied to a list with an even number of integers, the mean rounds half-way cases away from zero (as defined in the `AverageMean` reducer).

Reviewer 2: ```suggestion
**(4)**: If the median is applied to an array with an even number of entries, the mean rounds half-way cases away from zero (as defined in the `AverageMean` reducer).
```

> **(5)**: The previously `RadError::ModeEmpty` will be re-used and renamed as `RADError::EmptyArray`.

Reviewer 2: ```suggestion
**(5)**: The existing`ModeEmpty` error will be repurposed and renamed as `EmptyArray`.
```

> **(6)**: If the list of observations is empty, the RAD engine will output with an error as `RADError::EmptyArray`.

Reviewer 2: ```suggestion
**(6)**: If the array  is empty, the RAD engine will output an `EmptyArray` error.
```
